### PR TITLE
NSUrlSessionHandler should propagate network exceptions

### LINKED ISSUE #18458: [Bug] NSUrlSessionHandler doesn't propagate connection error correctly
With Xamarin iOS, the `HTTPClient `relying on a `NSUrlSessionHandler` may not consider connection loss as an error

This is related to https://github.com/xamarin/xamarin-macios/issues/5215#issuecomment-472955955

### Steps to Reproduce
1. Configure a `NSUrlSessionHandler` with a fairly short interval resource timeout
   a. In that case I am using a timeout exception, but other connection exception may end up with the same problem
3. Instantiate a `HTTPClient` using this handler
4. Connect to a HTTP server that slowly drips content
   a. The interval between data being sent should be greater than the interval resource timeout

> A unit test will be included in the PR for this issue, and can be used to reproduce the problem with the current `Xamarin.iOS` version

The underlying HTTP connection is terminated by the system as a network issue occurred.
The HTTP response content failed to being streamed entirely.

### Expected Behavior
The network exception is propagated through the `HTTPClient`

### Actual Behavior
The network exception is swallowed. The `HTTPClient` considers that the response content was successfully streamed which can cause problems down the line.

### Environment

<details>
<summary>Version information</summary>

```
Xamarin.iOS
Version: 16.4.0.6 Visual Studio Community
Hash: 97731c92c
Branch: xcode14.3
Build date: 2023-04-11 22:38:36-0400
```

</details>

### PR DESCRIPTION
**Pull request by EmilienDup**
fixes #18458

### CODE REVIEW

**Comment by EmilienDup** [comment]
@microsoft-github-policy-service agree company="SnagR Ltd."

File: tests/monotouch-test/System.Net.Http/MessageHandlers.cs [hunk 1]
```diff
@@ -271,6 +271,109 @@ public void TestNSUrlSessionEphemeralDisabledCookies ()
...
 
+		[Test]
+		public void TestNSUrlSessionTimeoutExceptionWhileStreamingContent ()
+		{
+			if (!HttpListener.IsSupported) {
+				Assert.Inconclusive ("HttpListener is not supported");
+			}
+
+			// HTPP listener config
+			IPEndPoint localEndPoint = new IPEndPoint (IPAddress.Any, 8080);
```
Comment (rolfbjarne) [inline tests/monotouch-test/System.Net.Http/MessageHandlers.cs:282]: Could the test use a dynamic port instead? In some cases we might end up executing the tests simulaneously on the same machine, and having a fixed port could cause conflicts if the same test executes in two different processes at thes ame time.

File: tests/monotouch-test/System.Net.Http/MessageHandlers.cs [hunk 2]
```diff
@@ -271,6 +271,109 @@ public void TestNSUrlSessionEphemeralDisabledCookies ()
...
+					timeoutExceptionWasThrown = true;
+				}
+			}, out var ex);
+
+			serverCancellationTokenSource.Cancel ();
+
+			if (!done) {
+				Assert.Inconclusive ("Test run timedout.");
+			} else if (!timeoutExceptionShouldHaveBeenThrown) {
+				Assert.Inconclusive ("Failed to produce a timeout. The response content was streamed completly.");
```
Comment (rolfbjarne) [inline tests/monotouch-test/System.Net.Http/MessageHandlers.cs:371]:
```suggestion
				Assert.Inconclusive ("Failed to produce a timeout. The response content was streamed completely.");
```

File: tests/monotouch-test/System.Net.Http/MessageHandlers.cs [hunk 3]
```diff
@@ -271,6 +271,109 @@ public void TestNSUrlSessionEphemeralDisabledCookies ()
...
+
+						timeoutExceptionWasThrown = false;
+						timeoutExceptionShouldHaveBeenThrown = outputStream.ToArray ().Length < expectedHttpResponseContentLength;
+					}
+				}
+				catch (Exception e)
+					when (e is TimeoutException || e is HttpRequestException) {
+					timeoutExceptionWasThrown = true;
+				}
+			}, out var ex);
```
Comment (rolfbjarne) [inline tests/monotouch-test/System.Net.Http/MessageHandlers.cs:364]: You're not doing anything with `ex`, maybe also add an assert that no exception was thrown:

```cs
Assert.NotNull (ex, "Exception");
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

**Review by mandel-macaque — APPROVED** [review]

**Comment by mandel-macaque** [comment]
@rolfbjarne @dalexsoto I believe the is a valid change and the failures in the CI are not related to this commit, but due to an update on the allowed words by Microsoft in the source code, specially in the mon source code:
```
 Tool: PoliCheck: Rule: 80034 (Term: nuke. TermClass: Cultural. Action: Action : Context = In all other contexts.; ActionRecommendation = <b>REMOVE</b>; Context = When used an an abbreviation of nuclear weapon.; ActionRecommendation = <b>Replace</b> with "nuclear weapon" or "nuclear armament"; ). 
  Term: 'nuke'. Term class: Cultural. Term table: English-9. Position: Line: 137. Context(term at 16): '  displayName: 'Nuke Provisionator Tool Cache''
```
And some nonsense related to the Mac APIs:
```
Tool: PoliCheck: Rule: 203588 (Term: Orient. TermClass: Geopolitical. Action: Action : Context = When used in other contexts, such as to refer to positioning.; ActionRecommendation = <b>Leave term unchanged</b>; Context = When used to refer to the geographic region of East Asia.; ActionRecommendation = <b>Replace</b> with "East Asia" or "Asia"; Context = When used to refer to the Microsoft sales region.; ActionRecommendation = <b>Replace</b> with "Asia"; ). 
  Term: 'Orient'. Term class: Geopolitical. Term table: English-9. Position: Line: 261626. Context(term at 34): 'P:MonoMac.WebKit.DomOverflowEvent.Orient'.
```

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**